pipeline/authz/remote_opa.go

```diff
@@ -0,0 +1,136 @@
+package authz
+
+import (
+	"bytes"
+	"encoding/json"
+	"net/http"
+	"strings"
+	"text/template"
+
+	"github.com/pkg/errors"
+
+	"github.com/ory/x/httpx"
+
+	"github.com/ory/oathkeeper/driver/configuration"
+	"github.com/ory/oathkeeper/helper"
+	"github.com/ory/oathkeeper/pipeline"
+	"github.com/ory/oathkeeper/pipeline/authn"
+	"github.com/ory/oathkeeper/x"
+)
+
+// AuthorizerRemoteOPAConfiguration represents a configuration for the remote_opa authorizer.
+type AuthorizerRemoteOPAConfiguration struct {
+	Remote string `json:"remote"`
+}
+
+// AuthorizerRemoteOPA implements the Authorizer interface.
+type AuthorizerRemoteOPA struct {
+	c configuration.Provider
+
+	client *http.Client
+	t      *template.Template
+}
+
+// NewAuthorizerRemoteOPA creates a new AuthorizerRemoteOPA.
+func NewAuthorizerRemoteOPA(c configuration.Provider) *AuthorizerRemoteOPA {
+	return &AuthorizerRemoteOPA{
+		c:      c,
+		client: httpx.NewResilientClientLatencyToleranceSmall(nil),
+		t:      x.NewTemplate("remote_opa"),
+	}
+}
+
+// OPA Response
+type OPAResponse struct {
+	Result struct {
+		Allow bool
+	}
+	Decision_id string
+}
+
+// OPA Policy Input
+type OPAInput struct {
+	User   string   `json:"user"`
```

Comment:
This payload is bound to your model of access control and is not generalizable for other use cases. Instead, this should be configurable (see e.g. `remote_json` for an implementation). You might also want to wait for JsonNet support (https://github.com/ory/oathkeeper/issues/441) for this.

Reply:
The payload for the OPA server needs the Method of the request, as well as the path of the request as an array. I believe I've added the Method correctly into the MatchContext structure, however I'm having a stab at the best way to add the Path "array" into the template.

Given it's specific to this module I don't think it makes sense to put it back in the AuthenticationSession struct but happy to take your guidance on that.

Reply:
> The payload for the OPA server needs the Method of the request, as well as the path of the request as an array. I believe I've added the Method correctly into the MatchContext structure, however I'm having a stab at the best way to add the Path "array" into the template.

I don't think that is correct. You define the input model using Rego, so it depends on your Rego Policies what input you expect. You're probably referring to [this example](https://www.openpolicyagent.org/docs/latest/http-api-authorization/#2-load-a-policy-into-opa) but as you can see from the Rego policy, the input parameters are defined by the policy itself, not by OPA.

Reply:
Absolutely, the intention though was for it to be able to work with the OPA tutorial you linked "out of the box" and I've also seen this "path split into an array" in use by other developers I'm working with.

Any suggestions on how this could be achieved any other way using text/template, or do you have any objections to this approach?

Reply:
I think the best approach would be using JsonNet (see https://github.com/ory/oathkeeper/issues/441) but I feat that needs quite some work on my end beforehand.

Reply:
Looks like there's no way to implement this in an acceptable way without JsonNet so I'll close off this PR and look at it again at some point in the future